File: ProblemList/00072.py
# DP
# class Solution:
#     def minDistance(self, word1: str, word2: str) -> int:
#         minimum_opers = []
#         word1_len = len(word1)+1
#         word2_len = len(word2)+1
#         for _ in range(word1_len):
#             minimum_opers.append([0 for _ in range(word2_len)])
#         # minimum_opers[i][j]: means that use minimum_opers[i][j] operations to make word1[:i] equals to word2[:j]
            
#         for i in range(word1_len):
#             for j in range(word2_len):
#                 if i == 0:
#                     minimum_opers[i][j] = j
#                 elif j == 0:
#                     minimum_opers[i][j] = i
#                 # make word1[:i] == word2[:j], we can choose minimum of three operations:
#                 # insert: needs minimum_opers[i][j-1]+1 operations
#                 # delete: needs minimum_opers[i-1][j]+1 operations
#                 # replace: 
#                 #   if word1[i-1] == word2[j-1] needs minimum_opers[i-1][j-1] operations
#                 #   else needs minimum_opers[i-1][j-1]+1 operations
#                 else:
#                     replace = minimum_opers[i-1][j-1] if word1[i-1] == word2[j-1] else minimum_opers[i-1][j-1]+1
#                     minimum_opers[i][j] = min(minimum_opers[i][j-1]+1,minimum_opers[i-1][j]+1,replace)

#         return minimum_opers[word1_len-1][word2_len-1]

class Solution:
    def minDistance(self, word1: str, word2: str) -> int:
        dp = [[0 for _ in range(len(word2)+1)] for _ in range(len(word1)+1)]

        sum_ = 1
        for i in range(len(word1)):
            dp[i+1][0] = sum_
            sum_ += 1
        
        sum_ = 1
        for i in range(len(word2)):
            dp[0][i+1] = sum_
            sum_ += 1

        for i in range(len(word1)):
            for j in range(len(word2)):
                if word1[i] == word2[j]: dp[i+1][j+1] = dp[i][j]
                else:
                    dp[i+1][j+1] = min(dp[i][j+1],dp[i+1][j],dp[i][j])+1 # delete, insert, replace 

        return dp[len(word1)][len(word2)]

# 1d
class Solution:
    def minDistance(self, word1: str, word2: str) -> int:
        dp = [0 for _ in range(len(word2)+1)]

        sum_ = 1
        for i in range(len(word2)):
            dp[i+1] = sum_
            sum_ += 1
        
        for i in range(len(word1)):
            cur = [dp[0]+1]
            for j in range(len(word2)):
                if word1[i] == word2[j]: cur.append(dp[j])
                else:
                    cur.append(min(dp[j+1],cur[j],dp[j])+1)
            dp = cur

        return dp[-1]

if __name__ == '__main__':
    a = Solution()
    print(a.minDistance("horse","ros"))
    print(a.minDistance("intention","execution"))

File: ProblemList/00072.cpp
#include<string>
#include <vector>
using namespace std;

// tabulation
class Solution {
public:
    int minDistance(string word1, string word2) {
        vector<vector<int>> dp(word1.size()+1, vector<int>(word2.size()+1,INT_MAX));

        dp[0][0] = 0;
        for(int i=1; i<=word1.size(); i++){
            dp[i][0] = dp[i-1][0] + 1;
        }

        for(int i=1; i<=word2.size(); i++){
            dp[0][i] = dp[0][i-1] + 1;
        }
        for(int i=1; i<=word1.size(); i++){
            for(int j=1; j<=word2.size(); j++){
                if(word1[i-1] == word2[j-1]) dp[i][j] = dp[i-1][j-1];
                else dp[i][j] = min(dp[i-1][j],min(dp[i][j-1],dp[i-1][j-1])) + 1;
            }
        }
        return dp[word1.size()][word2.size()];
    }
};

// memorization
class Solution {
private:
    int inf = 10000000;
    int dp(int idx1, int idx2, string& s1, string& s2, vector<vector<int>>& mem){
        if(idx1 == s1.size() && idx2 == s2.size()) return 0;
        else if(idx1 > s1.size() || idx2 > s2.size()) return inf;
        else if(mem[idx1][idx2] != inf) return mem[idx1][idx2];

        int ret = inf;
        if(idx1 < s1.size() && idx2 < s2.size() && s1[idx1] == s2[idx2]) ret = min(ret, dp(idx1+1, idx2+1, s1, s2, mem));
        else{
            // insert
            ret = min(ret, 1+dp(idx1, idx2+1, s1, s2, mem));
            // delete
            ret = min(ret, 1+dp(idx1+1, idx2, s1, s2, mem));
            // replace
            ret = min(ret, 1+dp(idx1+1, idx2+1, s1, s2, mem));
        }
        mem[idx1][idx2] = ret;
        return ret;
    }
public:
    int minDistance(string word1, string word2) {
        vector<vector<int>> mem(word1.size()+1, vector<int>(word2.size()+1,inf));
        return dp(0, 0, word1, word2, mem);
    }
};